### ISSUE TITLE
Issue with list slice operation

### ISSUE THREAD

**Issue by UdayaR** [enthought/traits#283]
When i try to execute the below code 
`from traits.api import HasTraits, List
class Sample(HasTraits):
    a = List
    
s = Sample(a = [1,2,3,4,5,6])
s.a[1:5:2] = [2,2,2]`
I get the following error

> ValueError: attempt to assign sequence of size 3 to extended slice of size 2

**Comment by mdickinson** [comment]
This is nothing to do with Traits: the error comes directly from Python:
```
>>> a = [1, 2, 3, 4, 5, 6]
>>> a[1:5:2] = [2, 2, 2]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: attempt to assign sequence of size 3 to extended slice of size 2
```
The expression `a[1:5:2]` describes a 2-element sublist of `a`, so it's not meaningful to assign 3 elements to it: where should those 3 elements be placed?

**Comment by UdayaR** [comment]
Sorry wrong example. Please see the below example
from traits.api import HasTraits, List
class Sample(HasTraits):
    a = List

a = range(6)
a[1:6:2] = [2,2,2]
print 'python syntax works'
s = Sample(a = range(6))
s.a[1:6:2] = [2,2,2]
print 'traits works'

**Comment by mdickinson** [comment]
Ah, yes. Indeed that looks wrong (and I can reproduce - see below); thanks for the report!

Without having looked at the code yet, this *looks* like a simple off-by-one error. If that's the case, it should be an easy fix ...

```python
>>> from traits.api import *
>>> class Sample(HasTraits):
...     a = List
... 
>>> s = Sample(a = range(6))
>>> s.a
[0, 1, 2, 3, 4, 5]
>>> s.a[1:6:2]
[1, 3, 5]
>>> s.a[1:6:2] = [2, 2, 2]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "build/bdist.macosx-10.6-x86_64/egg/traits/trait_handlers.py", line 2361, in __setitem__
ValueError: attempt to assign sequence of size 3 to extended slice of size 2
```

**Comment by UdayaR** [comment]
I also located the issue in the traits library
In trait_handlers.py line number 2356
`slice_len = max(0, (key.stop - key.start) // key.step)`
to be replaced with 
`slice_len = max(0, math.ceil(float(key.stop - key.start) / key.step))``

**Comment by UdayaR** [comment]
Thanks for your help..!

**Comment by mdickinson** [comment]
Thanks, that saved me some time. I agree that that's the right fix, though I'd suggest avoiding floating-point arithmetic and sticking to integer operations:
```python
slice_len = max(0, -((key.start - key.stop) / key.step))
```
And that should work for both positive and negative steps.

Now we just need some unit tests!


  I think the right formula here is `max(0, - ((key.start - key.stop) // abs(key.step)))`, which avoids the use of floating-point arithmetic and also works for negative steps.

We need some unittests ...

**Comment by UdayaR** [comment]
Yes, this solution looks elegant

**Comment by mdickinson** [comment]
There's another more subtle bug here, namely that the slicing code is using `slice(*key.indices(len( self )))` as a new valid slice index. That doesn't work. (See http://bugs.python.org/issue1256786.)

Also, `__delitem__` needs a similar off-by-one fix.

**Comment by mdickinson** [comment]
Fixed in #285.
